README.md
# Anima Utilities

Command line utilities for playing the [Anima: Beyond Fantasy](https://tvtropes.org/pmwiki/pmwiki.php/TabletopGame/AnimaBeyondFantasy) tabletop game.

Note that these calculations are not guaranteed to be from either the Core rulebook or the updated version, Core Exxet. I believe that I'm using the damage formula from Core Exxet. If you see a problem, [open an issue](https://github.com/desmeraldoo/anima/issues/new).

## Installation

### From Pypi

1. Install [Python](https://www.python.org/) and [Git](https://git-scm.com/) if not already installed.
1. Run `pip install anima-utils` to install the program.
1. Run the command `anima` to open the Anima Utils CLI.

### From Source

1. Use Git to [clone](https://docs.github.com/en/free-pro-team@latest/github/creating-cloning-and-archiving-repositories/cloning-a-repository) the [repo](https://github.com/desmeraldoo/anima).
1. Launch a [command prompt](https://www.howtogeek.com/235101/10-ways-to-open-the-command-prompt-in-windows-10/) in the repository folder, `anima`.
1. Run `python install .`, which will build and install the package and add the `anima` command to your system.
1. Run `anima` to open the command prompt.

If you have any problems following the steps above, [open an issue](https://github.com/desmeraldoo/anima/issues/new).

## Documentation

The below demonstrations of how to use the commands use examples from the *Anima: Beyond Fantasy* core book or original illustrative examples.

### atk

Resolves an attack and defense exchange.

`usage: atk attack_roll defense_roll armor_value base_damage`

| field        | type | default | description                                                               |
| ------------ | ---- | ------- | ------------------------------------------------------------------------- |
| attack_roll  | int  | 0       | The final result of the attacker's offensive roll.                        |
| defense_roll | int  | 0       | The final result of the defender's defensive roll.                        |
| attack_roll  | int  | 0       | The defender's Armor on the Table corresponding to the attacker's attack. |
| attack_roll  | int  | 0       | The base damage of the weapon used by the attacker, if known or relevant. |

> For example, Celia attacks one of the guards with whom she was earlier locked in combat. Celia now has an Attack Ability of 120, while the guard’s Dodge is only 60. Both of them roll the dice. Celia rolls an 86, which, added to her Ability, gives her a Final Attack of 206. The guard’s dice roll is a 44, and so his Final Defense is 104. As Celia is the attacker, the guard’s Final Defense is subtracted from Celia’s Final Attack (206 – 104). The result used when referencing Table 38 is, therefore, 102.

```
(anima) attack 206 104
ATTACK: 100%
```

> Celia had just made an attack with a favorable margin of 102 against the city guard. Since she is wielding a saber against an enemy wearing hardened leather armor, her opponent uses an AT of 2 against her Cutting attack. Consulting the Combat Table, we see that Celia’s attack produces 80% damage. Since the Final Damage number is 45 (only the saber’s Base Damage in this case, since Celia has a 0 Strength Bonus), the guard suffers only 36 points of damage (80% of 45).

```
(anima) attack 206 104 2 45
DAMAGE: 36
```

### combo

Enters a special interactive mode where the user can specify multiple attacks and resolve them all at once.

There is no need to enter the total number of attacks up front. The longest list is taken as the true number of attacks. Blanks are replaced with zero, and `'.'` values are replaced with the closest preceding non-null value, or zero, in the absence of any such value.

`usage: combo`

```
(anima) combo
atks:   150 180 170 210
defs:   160 120 200 130
amrs:   2 4 6 8
dmgs:   100 80 60 40

╒══════╤══════════╤═══════════╤═════════╤═══════════════╤══════════════════════╤══════════╕
│   id │   attack │   defense │   armor │   base_damage │ result               │   damage │
╞══════╪══════════╪═══════════╪═════════╪═══════════════╪══════════════════════╪══════════╡
│    0 │      150 │       160 │       2 │           100 │ COUNTERATTACK: +5 C  │        0 │
├──────┼──────────┼───────────┼─────────┼───────────────┼──────────────────────┼──────────┤
│    1 │      180 │       120 │       4 │            80 │ DAMAGE: 16           │       16 │
├──────┼──────────┼───────────┼─────────┼───────────────┼──────────────────────┼──────────┤
│    2 │      170 │       200 │       6 │            60 │ COUNTERATTACK: +15 C │        0 │
├──────┼──────────┼───────────┼─────────┼───────────────┼──────────────────────┼──────────┤
│    3 │      210 │       130 │       8 │            40 │ MISSED               │        0 │
╘══════╧══════════╧═══════════╧═════════╧═══════════════╧══════════════════════╧══════════╛
╒═════════╕
│   Total │
╞═════════╡
│      16 │
╘═════════╛
```

```
(anima) combo
atks:   150 160 170 180
defs:
amrs:   10
dmgs:   50 . . 150

╒══════╤══════════╤═══════════╤═════════╤═══════════════╤═════════════╤══════════╕
│   id │   attack │   defense │   armor │   base_damage │ result      │   damage │
╞══════╪══════════╪═══════════╪═════════╪═══════════════╪═════════════╪══════════╡
│    0 │      150 │         0 │      10 │            50 │ DAMAGE: 25  │       25 │
├──────┼──────────┼───────────┼─────────┼───────────────┼─────────────┼──────────┤
│    1 │      160 │         0 │      10 │            50 │ DAMAGE: 30  │       30 │
├──────┼──────────┼───────────┼─────────┼───────────────┼─────────────┼──────────┤
│    2 │      170 │         0 │      10 │            50 │ DAMAGE: 35  │       35 │
├──────┼──────────┼───────────┼─────────┼───────────────┼─────────────┼──────────┤
│    3 │      180 │         0 │      10 │           150 │ DAMAGE: 120 │      120 │
╘══════╧══════════╧═══════════╧═════════╧═══════════════╧═════════════╧══════════╛
╒═════════╕
│   Total │
╞═════════╡
│     210 │
╘═════════╛
```

```
(anima) combo
atks:   280
defs:   140 180 320 150
amrs:   . . . 8
dmgs:   100

╒══════╤══════════╤═══════════╤═════════╤═══════════════╤══════════════════════╤══════════╕
│   id │   attack │   defense │   armor │   base_damage │ result               │   damage │
╞══════╪══════════╪═══════════╪═════════╪═══════════════╪══════════════════════╪══════════╡
│    0 │      280 │       140 │       0 │           100 │ DAMAGE: 140          │      140 │
├──────┼──────────┼───────────┼─────────┼───────────────┼──────────────────────┼──────────┤
│    1 │      280 │       180 │       0 │           100 │ DAMAGE: 100          │      100 │
├──────┼──────────┼───────────┼─────────┼───────────────┼──────────────────────┼──────────┤
│    2 │      280 │       320 │       0 │           100 │ COUNTERATTACK: +20 C │        0 │
├──────┼──────────┼───────────┼─────────┼───────────────┼──────────────────────┼──────────┤
│    3 │      280 │       150 │       8 │           100 │ DAMAGE: 50           │       50 │
╘══════╧══════════╧═══════════╧═════════╧═══════════════╧══════════════════════╧══════════╛
╒═════════╕
│   Total │
╞═════════╡
│     290 │
╘═════════╛
```

### crit

Calculates the effects of a critical, or reports that a critical has no effects. If the critical involves amputation, the location is randomly chosen unless specified (by an integer value corresponding to Table 48 of the core book).

`usage: crit level phr_roll location_id`

| field       | type | default | description                                                                                                                                                                                                                                                                                   |
| ----------- | ---- | ------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| level       | int  | 0       | The damage dealt by the attack that triggered the critical roll, plus 1d100 rolled by the attacker. Note that if this value is over 200, the script automatically reduces the value in excess of 200 by half, per the rules for criticals given in the core book.                             |
| phr_roll    | int  | 0       | The final result of the victim's Physical Resistance roll.                                                                                                                                                                                                                                    |
| attack_roll | int  | 0       | The defender's Armor on the Table corresponding to the attacker's attack.                                                                                                                                                                                                                     |
| location_id | int  | 0       | A way of specifying the location of the critical, if for example the attacker was targeting a specific region of the body. See Table 48 on Page 90 of the core book to look up locations, or simply ignore the output of the program with respect to location if one has already been chosen. |

> Muris has just accosted a shopkeeper who was badgering him about his five-finger discount. The shopkeeper doesn't expect his swift sucker punch, so Muris succeeds in his attack and causes 50 damage. The merchant, having only 90 LP, suffers a critical blow, and two more rolls are required to determine the extent of the damage. Muris rolls 1d100 and adds it to the damage caused, and the merchant makes a Physical Resistance Check, using his PhR score of 30 as a base.
>
> Muris rolls 70, and the merchant rolls only 13, so this is a dire blow with a critical level of 77. The GM rolls to determine the location of the critical, and their roll yields a 67, or the merchants left hand, which seems to have been badly broken in his failed attempt to block the strike.
>
> Thus, in addition to the damage he received, the merchant suffers not only a -38 All Action Penalty that diminishes at a rate of 5 points per round for the duration of the encounter, but an additional -38 All Action Penalty that heals at a rate dependent on his Constitution, normally 5 points per day (see the core book, page 53). At the discretion of the GM, he may require further medical attention to heal and regain use of his hand. Though, perhaps he should count himself fortunate in one respect—if Muris's critical roll been just 25 points higher, the hand would have been crushed beyond repair.

```
(anima) crit 120 30
CRIT LEVEL 90
        MAJOR CRITICAL [Left arm (Hand)]
        ALL ACTION PENALTY (-45)
        ALL ACTION SACRIFICE (-45)
```


CHANGELOG.md
# Changelog

## 1.0.0

- Convert to Pleroma toolkit

## 0.3.2

- Fix `AttackResolver` not properly calculating damage

## 0.3.1

- Fix tests import error caused by a lack of attention to detail

## 0.3.0

- Rename the `chaos` command to the `powers` command
- Add powers from the book "Those Who Walked Amongst Us" to the `powers` command
- Fix `combo` algorithm boolean logic
- Fix `combo` posting a stacktrace on keyboard interrupt

## 0.2.0

- Remove auto-cap of Armor Value to accommodate Armor Values above 10
- `atk` and its dependencies now only report `"MISSED"` if the attack missed due to Absorption
  - Previously, it would report `"MISSED"` if the damage was perfectly nullfied by Armor without being reduced into the negative range

## 0.1.4

- Correct method of adding `tabulate` to dependencies in `pyproject.toml`

## 0.1.3

- Update project name to `anima-utils` to avoid conflict on PyPI

## 0.1.2

- Add `tabulate` to dependencies in `pyproject.toml`

## 0.1.1

- Update [documentation](README.md)

## 0.1.0

- Refurbish combo command to help with complicated multiattack scenarios

## 0.0.0

- Restructure repo into package form
- Add pre-commit hooks
- Introduce changelog


anima-utils.py
import cmd
import csv
import math
import random
import sys
import pdb

DEFAULT_PARTY_FILE = 'my-campaign.csv'
SHOCK_CONSEQUENCES = [
    '\n\t-1 Mental Health',
    '\n\t-2 Mental Health',
    '\n\t-3 Mental Health',
    '\n\t-5 Mental Health',
    '\n\t-8 Mental Health',
    '\n\t-10 Mental Health',
    '\n\t-15 Mental Health\n\tPHYSICAL SHOCK 60',
    '\n\t-20 Mental Health\n\tPHYSICAL SHOCK 80\n\tMINOR TEMPORARY DERANGEMENT',
    '\n\t-30 Mental Health\n\tPHYSICAL SHOCK 100\n\tMINOR TEMPORARY DERANGEMENT',
    '\n\t-40 Mental Health\n\tPHYSICAL SHOCK 120\n\tMAJOR TEMOPRARY DERANGEMENT',
    '\n\t-50 Mental Health\n\tPHYSICAL SHOCK 140\n\tMAJOR TEMPORARY DERANGEMENT'
]
POWERS = [
    'Fatigue Resistance',
    'Acute Senses',
    'Attribute Increased',
    'Unnatural Size',
    'Inhumanity/Zen',
    'Aquatic Breathing',
    'Reduced Physical Needs',
    'Natural Immunity to an Element',
    'Psychological Immunity',
    'Natural Weapons',
    'Additional Attacks',
    'Increased Damage',
    'Increased Reaction',
    'Damage Energy',
    'Armor Modifier',
    'Special Attack',
    'Poisoned Attack',
    'Added Mystical Effect',
    'Increased Critical',
    'Special Trapping',
    'Supernatural Attack',
    'Elemental Attack',
    'Special Movement',
    'Automatic Transport',
    'Increased Movement',
    'Natural Flight',
    'Mystical Flight',
    'Increased Physical Resistance',
    'Mystical/Psychic Resistance',
    'Regeneration',
    'Physical Immunity',
    'Magical Immunity',
    'Matrix Immunity',
    'Damage Barrier',
    'Physical Armor',
    'Mystical Armor',
    'Innate Magic',
    'Innate Psychic Abilities',
    'Metamorphosis',
    'Invisibility / Undetectable',
    'Aura',
    'Special Means of Vision',
    'Supernatural Detection'
]

def parse(args):
    '''Convert a series of zero or more numbers to an argument tuple'''
    return tuple(map(int, args.split()))

def roll(score, modifier=0, open_roll=0, name=''):
    base_roll = random.randint(1, 100)
    if open_roll > -1 and (base_roll >= (90 + open_roll) or base_roll == 100):
        print('OPEN ROLL!' if not name else f'OPEN ROLL: {name}')
        return score + base_roll + modifier + roll(0, open_roll=open_roll + 1, name=name)
    elif (open_roll < 1) and ((score < 200 and base_roll < 4) or (base_roll < 3)):
        print('FUMBLE...' if not name else f'FUMBLE: {name}')
        return score + base_roll + modifier - random.randint(1, 100)
    else:
        return score + base_roll + modifier

def calc_attack(attack, defense, armor=0, base_damage=None):
    result = attack - defense
    if result > 29:
        damage_multiple = None
        
        if armor < 2 and result < 50:
            # special rules for absorption
            if armor == 1:
                if result < 40:
                    damage_multiple = 0.1
                else:
                    damage_multiple = 0.3
            else:
                if result < 40:
                    damage_multiple = 0.1
                else:
                    damage_multiple = 0.2
        else:
            damage_multiple = ((result // 10) / 10) - (armor * 0.1)
        if damage_multiple <= 0:
            return 'DEFLECTED'
        elif not base_damage:
            return f'ATTACK: {int(damage_multiple * 100)}%'
        else:
            return f'DAMAGE: {int(round(base_damage * damage_multiple, 0))}'
    elif result < 0:
        counterattack_bonus = ((round(abs(result), -1) // 2)) - 5
        return f'COUNTERATTACK: +{counterattack_bonus} C'
    else:
        return 'MISSED'

def calc_crit(damage_dealt, phr_roll, modifier=0, location_level=0):
    crit_level = damage_dealt + random.randint(1, 100) + modifier
    if crit_level > 200: crit_level -= (crit_level - 200) // 2
    crit_level -= phr_roll
    
    if not location_level: location_level = random.randint(1, 100)
    location = ''
    if location_level < 11:
        location = 'Torso (Ribs)'
    elif location_level < 21:
        location = 'Torso (Shoulder)'
    elif location_level < 31:
        location = 'Torso (Stomach)'
    elif location_level < 36:
        location = 'Torso (Kidneys)'
    elif location_level < 49:
        location = 'Torso (Chest)'
    elif location_level < 51:
        location = 'Torso (Heart)'
    elif location_level < 55:
        location = 'Right arm (Upper forearm)'
    elif location_level < 59:
        location = 'Right arm (Lower forearm)'
    elif location_level < 61:
        location = 'Right arm (Hand)'
    elif location_level < 65:
        location = 'Left arm (Upper forearm)'
    elif location_level < 69:
        location = 'Left arm (Lower forearm)'
    elif location_level < 71:
        location = 'Left arm (Hand)'
    elif location_level < 75:
        location = 'Right leg (Thigh)'
    elif location_level < 79:
        location = 'Right leg (Calf)'
    elif location_level < 81:
        location = 'Right leg (Foot)'
    elif location_level < 85:
        location = 'Left leg (Thigh)'
    elif location_level < 89:
        location = 'Left leg (Calf)'
    elif location_level < 91:
        location = 'Left leg (Foot)'
    else:
        location = 'Head'
    
    s = f'CRIT LEVEL {crit_level}'    
    if crit_level < 1:
        return s + '\n\tNO FURTHER EFFECT'
    elif crit_level < 51:
        return s + f'\n\tMINOR CRITICAL\n\tALL ACTION PENALTY (-{crit_level})'
    else:
        s += f'\n\tMAJOR CRITICAL [{location}]\n\tALL ACTION PENALTY (-{crit_level // 2})\n\tALL ACTION SACRIFICE (-{crit_level // 2})'
        
        if crit_level > 100:
            if location_level > 90 or (location_level == 49 or location_level == 50):
                s += '\n\tINSTANT DEATH'
            elif location_level > 50 and location_level < 91:
                s += '\n\tAMPUTATION'
        
            if crit_level > 150:
                s += '\n\tUNCONCIOUS -- AT RISK OF DEATH'
        
        return s

def calc_shock(composure, willpower, difficulty=10, name=''):
    # This is a homebrew difficulty calculation that does not reflect the
    # system as described in the Master's Tookit, which I found inconsistent. I
    # tried to make it as similar as possible.
    composure_difficulty = 20 + (difficulty * 10)
    composure_roll = roll(composure, name=name)
    s = f'Composure roll {composure_roll} against difficulty {composure_difficulty}'
    if composure_roll > composure_difficulty:
        return s + '\n\tRESISTED: COMPOSURE'
    
    shock_level = (willpower + random.randint(1, 10)) - (difficulty)
    s += f'\nLevel of failure: {shock_level}'
    if shock_level > -1:
        return s + '\n\tRESISTED: WILLPOWER'
    elif shock_level > -11:
        return s + SHOCK_CONSEQUENCES[abs(shock_level) - 1]
    else:
        return s + SHOCK_CONSEQUENCES[-1]

def group_shock(difficulty=10):
    with open(DEFAULT_PARTY_FILE) as csvfile:
        reader = csv.DictReader(csvfile)
        s = []
        for row in reader:
            s.append(row['name'])
            s.append(calc_shock(int(row['composure']), int(row['willpower']), difficulty, row['name']))
    return '\n'.join(s)

def calc_notice(score, difficulty, modifier=0, name=''):
    notice_score = roll(score, modifier, name=name)
    notice_result = 'noticed' if notice_score >= difficulty else 'missed'
    return f'{notice_result} (score: {notice_score}, difficulty: {difficulty})'
    
def group_notice(difficulty, modifier=0):
    with open(DEFAULT_PARTY_FILE) as csvfile:
        reader = csv.DictReader(csvfile)
        s = []
        for row in reader:
            s.append(row['name'])
            s.append(': ')
            s.append(calc_notice(int(row['notice']), difficulty, modifier, row['name']))
            s.append('\n')
    return ' '.join(s)

def calc_mr(score, difficulty, modifier=0, name=''):
    resistance = roll(score, modifier, open_roll=-1, name=name)
    resist_result = 'RESISTED' if resistance >= difficulty else 'AFFECTED'
    return f'{resist_result} (score: {resistance}, difficulty: {difficulty})' 

def group_mr(difficulty=80, modifier=0):
    with open(DEFAULT_PARTY_FILE) as csvfile:
        reader = csv.DictReader(csvfile)
        s = []
        for row in reader:
            s.append(row['name'])
            s.append(calc_mr(int(row['magicresist']), difficulty, modifier, row['name']))
    return '\n'.join(s)

def chaotic_powers(num_powers=5):
    return 'CHAOTIC POWERS:\n' + '\n'.join(random.choices(POWERS, k=num_powers))

class AnimaShell(cmd.Cmd):
    intro = 'Engaging Anima toolkit'
    prompt = '(anima) '
    
    def default(self, _):
        print('Input not recognized.')
    
    def do_attack(self, args):
        'usage: attack attack_roll defense_roll [armor] [base_damage]'
        print(calc_attack(*parse(args)))
    
    def do_crit(self, args):
        'usage: crit damage_dealt phr_roll [modifier] [location_level]'
        print(calc_crit(*parse(args)))
    
    def do_shock(self, args):
        'usage: shock composure willpower [difficulty=10]'
        print(calc_shock(*parse(args)))
    
    def do_groupshock(self, args):
        'usage: groupshock [difficulty=10]\nrequires user-defined file \'my-campaign.csv\' (see docs)'
        print(group_shock(*parse(args)))
    
    def do_notice(self, args):
        'usage: notice score difficulty [modifier=0]'
        print(calc_notice(*parse(args)))
    
    def do_groupnotice(self, args):
        'usage: groupnotice difficulty [modifier=0]\nrequires user-defined file \'my-campaign.csv\' (see docs)'
        print(group_notice(*parse(args)))
    
    def do_groupmr(self, args):
        'usage: groupmr difficulty [modifier=0]\nrequires user-defined file \'my-campaign.csv\' (see docs)'
        print(group_mr(*parse(args)))
    
    def do_chaos(self, args):
        'usage: chaos [num_powers]'
        print(chaotic_powers(*parse(args)))
        
    def do_exit(self, args):
        'terminate session'
        print('Ending session')
        return True

if __name__ == '__main__':
    AnimaShell().cmdloop()